### PR TITLE
Slasher highest source target

### PR DESCRIPTION
**Pull request by alonmuroch**
Please see discussion: https://github.com/prysmaticlabs/prysm/pull/7241

For some reason GIthub marked the previous PR as unknown repository.

### CODE REVIEW

**Comment by shayzluf** [comment]
LGTM @farazdagi please take a look as well

File: slasher/cache/highest_attestation_cache_test.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	require.NoError(t, err)
+
+	// cache
+	cache.Set(1, &ethereum_slashing.HighestAttestation{
+		ValidatorId:        1,
+		HighestSourceEpoch: 2,
+		HighestTargetEpoch: 3,
+	})
+
+	// has
```
Comment (farazdagi) [inline slasher/cache/highest_attestation_cache_test.go:22]: Full english sentences as comments are preferred. Adding comments as delimiters is probably not a good idea, comments should explain sth non-trivial. Such laconic comments serve little purpose explaining things (they are too short for that). Even in tests, let's remove obvious comments (for example `// delete` followed by delete operation serves no purpose).

File: slasher/cache/highest_attestation_cache_test.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	})
+	cache.Set(3, &ethereum_slashing.HighestAttestation{
+		ValidatorId:        7,
+		HighestSourceEpoch: 8,
+		HighestTargetEpoch: 9,
+	})
+
+	// purge
+	cache.Purge()
+
```
Comment (farazdagi) [inline slasher/cache/highest_attestation_cache_test.go:67]:
```suggestion
```

File: slasher/cache/highest_attestation_cache_test.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	require.Equal(t, false, cache.Has(1))
+	require.Equal(t, false, cache.Has(2))
+	require.Equal(t, false, cache.Has(3))
+
```
Comment (farazdagi) [inline slasher/cache/highest_attestation_cache_test.go:71]:
```suggestion
```

File: slasher/cache/highest_attestation_cache_test.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	})
+	cache.Set(3, &ethereum_slashing.HighestAttestation{
+		ValidatorId:        7,
+		HighestSourceEpoch: 8,
+		HighestTargetEpoch: 9,
+	})
+
+	// purge
+	cache.Clear()
+
```
Comment (farazdagi) [inline slasher/cache/highest_attestation_cache_test.go:102]:
```suggestion
```

File: slasher/cache/highest_attestation_cache_test.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	require.Equal(t, false, cache.Has(1))
+	require.Equal(t, false, cache.Has(2))
+	require.Equal(t, false, cache.Has(3))
+
```
Comment (farazdagi) [inline slasher/cache/highest_attestation_cache_test.go:106]:
```suggestion
```

File: slasher/db/kv/highest_attestation.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// EnableHighestAttestationCache used to enable or disable highest attestation cache in tests.
+func (db *Store) EnableHighestAttestationCache(enable bool) {
+	db.highestAttCacheEnabled = enable
+}
+
+// HighestAttestation returns the highest calculated attestation for a validatorID
+func (db *Store) HighestAttestation(ctx context.Context, validatorID uint64) (*slashpb.HighestAttestation, error) {
+	ctx, span := trace.StartSpan(ctx, "slasherDB.HighestAttestation")
```
Comment (farazdagi) [inline slasher/db/kv/highest_attestation.go:47]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "SlasherDB.HighestAttestation")
```

File: slasher/db/kv/highest_attestation.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+		}
+		return nil
+	})
+
+	return highestAtt, err
+}
+
+// SaveHighestAttestation saves highest attestation for a validatorID
+func (db *Store) SaveHighestAttestation(ctx context.Context, highest *slashpb.HighestAttestation) error {
+	ctx, span := trace.StartSpan(ctx, "SlasherDB.HighestAttestation")
```
Comment (farazdagi) [inline slasher/db/kv/highest_attestation.go:77]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "SlasherDB. SaveHighestAttestation")
```

File: slasher/db/kv/highest_attestation.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+			}
+			highestAtt = set[validatorID]
+		}
+		return nil
+	})
+
+	return highestAtt, err
+}
+
+// SaveHighestAttestation saves highest attestation for a validatorID
```
Comment (farazdagi) [inline slasher/db/kv/highest_attestation.go:75]:
```suggestion
// SaveHighestAttestation saves highest attestation for a validatorID.
```

File: slasher/db/kv/highest_attestation.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	return nil
+}
+
+func highestAttSetkeyBytes(validatorID uint64) []byte {
+	return bytesutil.Uint64ToBytesBigEndian(highestAttSetkey(validatorID))
+}
+
+// divide validators by id into 1k-ish buckets (0-1000,1001-1999, etc)
```
Comment (farazdagi) [inline slasher/db/kv/highest_attestation.go:124]:
```suggestion
// divide validators by id into 1k-ish buckets (0-1000,1001-1999, etc).
```

File: slasher/detection/detect.go [hunk 10]
```diff
@@ -213,3 +214,38 @@ func isSurrounding(incomingAtt, prevAtt *ethpb.IndexedAttestation) bool {
 	return incomingAtt.Data.Source.Epoch < prevAtt.Data.Source.Epoch &&
 		incomingAtt.Data.Target.Epoch > prevAtt.Data.Target.Epoch
 }
+
+// UpdateHighestAttestation updates to the db the highest source and target attestations for a each validator
```
Comment (farazdagi) [inline slasher/detection/detect.go:218]:
```suggestion
// UpdateHighestAttestation updates to the db the highest source and target attestations for a each validator.
```

File: slasher/detection/service.go [hunk 11]
```diff
@@ -196,6 +196,10 @@ func (ds *Service) detectHistoricalChainData(ctx context.Context) {
 				}
 			}
 			ds.submitAttesterSlashings(ctx, slashings)
+
+			if err := ds.UpdateHighestAttestation(ctx, att); err != nil {
+				log.WithError(err).Errorf("could not update highest attestation")
```
Comment (farazdagi) [inline slasher/detection/service.go:201]:
```suggestion
				log.WithError(err).Error("Could not update highest attestation")
```

**Review by farazdagi — COMMENTED** [review]
Haven't spotted anything significant, just several minor fixes needed.

**Comment by alonmuroch** [comment]
@farazdagi fixed.

**Review by shayzluf — APPROVED** [review]
lgtm

**Pull request merged**